pkg/storage/resource_limiter.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package storage
+
+import (
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+)
+
+// ResourceLimiterOptions is defining limits for resource limiter to restrict number
+// of iterations.
+type ResourceLimiterOptions struct {
+	MaxRunTime time.Duration
+}
+
+// ResourceLimitReached indicates which resource threshold is exceeded.
+// Soft threshold is advisory as to stop building results at convenient
+// point, hard threshold is when iteration should stop straight away.
+type ResourceLimitReached int
+
+//go:generate stringer -type ResourceLimitReached
+const (
+	ResourceLimitNotReached  ResourceLimitReached = 0
+	ResourceLimitReachedSoft ResourceLimitReached = 1
+	ResourceLimitReachedHard ResourceLimitReached = 2
```

Comment:
Why do we need the soft and hard limits? If we simply have a limit, once the caller knows the limit is exceeded it knows it needs to stop as soon as viable (given the callers internal invariants).

Reply:
This is in line with what we do for sstable size. Callers would prefer to have chunks split by key, in that case they can process them immediately. But if we hit hard limit we stop asap and then caller will have to deal with only a part of versions for the key and combine them with the following chunk.